Error parsing error response for `POST /_matrix/client/v3/keys/signatures/upload`

My client is stuck in a loop:

```
rust-crypto: successfully made HTTP request: POST /_matrix/client/v3/keys/signatures/upload
18:24:58.559 react_devtools_backend_compact.js:2367 Error processing outgoing-message requests from rust crypto-sdk Error: deserialization failed: missing field `error` at line 1 column 128
    at push.1685.module.exports.__wbindgen_error_new (matrix_sdk_crypto_js.js:6676:1)
    at <wasm_bindgen::JsError as core::convert::From<E>>::from::h66f6a8cb01617678 (01b16b46:0x313e37)
    at <matrix_sdk_crypto_js::responses::OwnedResponse as core::convert::TryFrom<(matrix_sdk_crypto_js::requests::RequestType,http::response::Response<alloc::vec::Vec<u8>>)>>::try_from::hde971358fe62519b (01b16b46:0x19b93b)
    at olmmachine_markRequestAsSent (01b16b46:0x2118e1)
    at OlmMachine.markRequestAsSent (matrix_sdk_crypto_js.js:3221:1)
    at OutgoingRequestProcessor.makeOutgoingRequest (OutgoingRequestProcessor.ts:99:35)
    at async RustCrypto.outgoingRequestLoop (rust-crypto.ts:876:21)
```

Synapse is returning a response like this:

```json
{
    "failures": {
        "@richvdh:sw1v.org": {
            "EOZDSWJVGZ": {
                "status": 400,
                "errcode": "M_INVALID_SIGNATURE",
                "message": "400: Invalid signature"
            }
        }
    }
}
```

There is no `error` property, and the [spec](https://spec.matrix.org/v1.7/client-server-api/#post_matrixclientv3keyssignaturesupload) does not specify one (though the example rather implies that it should be there).

See also https://github.com/matrix-org/matrix-spec/issues/1594.

--- a/crates/ruma-client-api/Cargo.toml
+++ b/crates/ruma-client-api/Cargo.toml
@@ -33,6 +33,10 @@ compat-unset-avatar = []
 # even if its value is an empty list.
 compat-get-3pids = []
 
+# Accept `message` as an alias for `error` in `upload_signatures::v3::Failure`,
+# since that's what Synapse sends.
+compat-upload-signatures = []
+
 unstable-exhaustive-types = ["ruma-common/unstable-exhaustive-types"]
 unstable-msc2666 = []
 unstable-msc2448 = []

--- a/crates/ruma-client-api/src/keys/upload_signatures.rs
+++ b/crates/ruma-client-api/src/keys/upload_signatures.rs
@@ -101,6 +101,7 @@ pub mod v3 {
         errcode: FailureErrorCode,
 
         /// Human-readable error message.
+        #[cfg_attr(feature = "compat-upload-signatures", serde(alias = "message"))]
         error: String,
     }
 
@@ -115,6 +116,33 @@ pub mod v3 {
 
         #[doc(hidden)]
         _Custom(PrivOwnedStr),
+    }
+
+    #[cfg(all(test, feature = "client", feature = "compat-upload-signatures"))]
+    mod tests {
+        use ruma_common::user_id;
+
+        use super::{FailureErrorCode, ResponseBody};
+
+        #[test]
+        fn deserialize_synapse_response() {
+            const JSON: &str = r#"{
+                "failures": {
+                    "@richvdh:sw1v.org": {
+                        "EOZDSWJVGZ": {
+                            "status": 400,
+                            "errcode": "M_INVALID_SIGNATURE",
+                            "message": "400: Invalid signature"
+                        }
+                    }
+                }
+            }"#;
+
+            let parsed: ResponseBody = serde_json::from_str(JSON).unwrap();
+            let failure = &parsed.failures[user_id!("@richvdh:sw1v.org")]["EOZDSWJVGZ"];
+            assert_eq!(failure.errcode, FailureErrorCode::InvalidSignature);
+            assert_eq!(failure.error, "400: Invalid signature");
+        }
     }
 }
 

--- a/crates/ruma/Cargo.toml
+++ b/crates/ruma/Cargo.toml
@@ -123,6 +123,9 @@ compat-unset-avatar = ["ruma-client-api?/compat-unset-avatar"]
 # Always serialize the threepids response field in `get_3pids::v3::Response`,
 # even if its value is an empty list.
 compat-get-3pids = ["ruma-client-api?/compat-get-3pids"]
+# Accept `message` as an alias for `error` in `upload_signatures::v3::Failure`,
+# since that's what Synapse sends.
+compat-upload-signatures = ["ruma-client-api?/compat-upload-signatures"]
 # Allow extra characters in signature IDs not allowed in the specification.
 compat-signature-id = ["ruma-signatures?/compat-signature-id"]
 
@@ -192,6 +195,7 @@ unstable-unspecified = [
 # Private feature, only used in test / benchmarking code
 __ci = [
     "full",
+    "compat-upload-signatures",
     "unstable-unspecified",
     "unstable-sanitize",
     "unstable-msc1767",

